contracts/swaps/connectors/SwapsImplSovrynSwap.sol

```diff
@@ -0,0 +1,176 @@
+pragma solidity 0.5.17;
+
+import "../../core/State.sol";
+import "../../feeds/IPriceFeeds.sol";
+import "../../openzeppelin/SafeERC20.sol";
+import "../ISwapsImpl.sol";
+import "./interfaces/ISovrynSwapNetwork.sol";
+import "./interfaces/IContractRegistry.sol";
+
+
+contract SwapsImplSovrynSwap is State, ISwapsImpl {
+    using SafeERC20 for IERC20;
+
+    // bytes32 contractName = hex"42616e636f724e6574776f726b"; // "SovrynSwapNetwork"
+
+    function getContractHexName(string memory source) public pure returns (bytes32 result) {
+        assembly {
+            result := mload(add(source, 32))
+        }
+    }
+
+    function getSovrynSwapNetworkContract() public view returns(ISovrynSwapNetwork){
+        // sovrynSwapContractRegistryAddress is part of State.sol and set in ProtocolSettings.sol
+        IContractRegistry contractRegistry = IContractRegistry(sovrynSwapContractRegistryAddress);
+        return ISovrynSwapNetwork(contractRegistry.addressOf(getContractHexName("SovrynSwapNetwork")));
+    }
+
+    /**
+     * swaps the source token for the destination token on the oracle based amm.
+     * on loan opening: minSourceTokenAmount = maxSourceTokenAmount and requiredDestTokenAmount = 0
+     *      -> swap the minSourceTokenAmount
+     * on loan rollover: (swap interest) minSourceTokenAmount = 0, maxSourceTokenAmount = complete collateral and requiredDestTokenAmount > 0
+     *      -> amount of required source tokens to swap is estimated (want to fill requiredDestTokenAmount, not more). maxSourceTokenAMount is not exceeded.
+     * on loan closure: minSourceTokenAmount <= maxSourceTokenAmount and requiredDestTokenAmount >= 0
+     *      -> same as on rollover. minimum amount is not considered at all.
+     * @param sourceTokenAddress the address of the source tokens
+     * @param destTokenAddress the address of the destination tokens
+     * @param receiverAddress the address to receive the swapped tokens
+     * @param returnToSenderAddress the address to return unspent tokens to (when called by the protocol, it's always the protocol contract)
+     * @param minSourceTokenAmount the minimum amount of source tokens to swapped (only considered if requiredDestTokens == 0)
+     * @param maxSourceTokenAmount the maximum amount of source tokens to swapped
+     * @param requiredDestTokenAmount the required amount of destination tokens
+     * **/
+
+    function internalSwap(
+        address sourceTokenAddress,
+        address destTokenAddress,
+        address receiverAddress,
+        address returnToSenderAddress,
+        uint256 minSourceTokenAmount,
+        uint256 maxSourceTokenAmount,
+        uint256 requiredDestTokenAmount)
+        public
```

Comment:
is this ok? The function name is `internal`

Reply:
that's confusing naming, but yes, it needs to be this way. I think, it's called internalSwap, because there's another contract SwapExternal.sol. They renamed it on the new version to dexSwap I think.